### PR TITLE
Shop

### PR DESCRIPTION
**Pull request by JonaLam**
# Description

Added a shop system
new shop scene which sells item

## Related issue(s)

#119 

## List of changes

- Changed the shop scene into something.
- Added shop item class which we place in the scene to sell the diffrent items

## Additional notes

How items are priced is not implemented in any meaningful way other than a set number on each of the shop items
the pool of items the shop can sell is just strings to their position in the files 
The inventory might make you unable to click anything, to fix you can make the inventory be a child of the inventorynode rather than the tree

### CODE REVIEW

**Comment by Turtyo** [comment]
I think we should remove the test on the naming convention, it's more bothersome than anything else

File: Cards/CardWorld.gd [hunk 1]
```diff
@@ -24,6 +24,13 @@ func init_card(in_card_data: CardBase) -> void:
 			card_cast_type = GlobalEnums.CardCastType.TARGET
 			break
 
+## takes money and increases price of shop removal, should only be used when this is wanted
+func remove_this_from_player_deck_with_price() -> void:
+	ShopManager.remove_card_from_player_deck_with_price(card_data)
```
Comment (Turtyo) [inline Cards/CardWorld.gd:29]: there is no check if we actually have the gold to do that right ?

File: Shop/shop_item_relic.gd [hunk 2]
```diff
@@ -0,0 +1,16 @@
+extends ShopItem
+
+var relic : Relic
+@export var relicImage : TextureButton
+@export var clickHandler : ClickHandler
+
+func _ready() -> void:
+	var rand : RandomNumberGenerator = RandomNumberGenerator.new()
```
Comment (Turtyo) [inline Shop/shop_item_relic.gd:8]: same

File: Shop/shop_item_torch.gd [hunk 3]
```diff
@@ -0,0 +1,33 @@
+extends ShopItem
+
+@export var clickHandler : ClickHandler
+var torchTexture : PackedScene = preload("res://Shop/tourchDisplay.tscn")
```
Comment (Turtyo) [inline Shop/shop_item_torch.gd:4]: why is this not `@export` ? also it's `tourchDisplay` it should be renamed `torchDisplay` :laughing:

File: Shop/shop_item_torch.gd [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	clickHandler.on_click.connect(on_shop_item_clicked)
+	
+	var startPos : Vector2 = startPosTransform.position
+	
+	for i in amount:
+		var instance : TextureRect = torchTexture.instantiate()
+		heldTorches.append(instance) 
+		instance.position = startPos
+		startPos += Vector2.RIGHT * 5
+		add_child(instance)
```
Comment (Turtyo) [inline Shop/shop_item_torch.gd:19]: a little comment to explain what you are doing here would be good

File: #Scenes/Events/shop/CardRemoval.gd [hunk 5]
```diff
@@ -0,0 +1,22 @@
+extends CardPileUISetter
```
Comment (Turtyo) [inline #Scenes/Events/shop/CardRemoval.gd:1]: This should be inside `SceneScripts` I think

**Review by Turtyo — CHANGES_REQUESTED** [review]
Checking the scene inside the game, the overlay on the top right is not correctly on screen. The `proceed` button should automatically open the map, or the map movement should be set to allowed directly when entering the shop (I think the second option is better)

Opening the inventory still prevents it from interacting with anything else or being closed, as mentioned in #136 
That makes it impossible for me to add gold and thus to test that buying things in the shop actually work, how did you test that btw ?

File: Shop/ShopItem.gd [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+func _ready() -> void:
+	priceLabel.text = str(cost)
+
+func on_shop_item_clicked() ->void:
+	if(InventoryManager.gold_component.can_afford(cost)):
+		queue_free()
+		InventoryManager.gold_component.lose_gold(cost)
+		_give_player_item()
+	else:
+		print("can't afford")
```
Comment (Tysterman74) [inline Shop/ShopItem.gd:19]: I'd say send something to the scene here to inform the user that they don't have enough money to purchase the item. Probably a debug variable in DEBUG_VAR to let us purchase w/o having to have gold, for testing purposes!

**Comment by Tysterman74** [comment]
Also, if we're removing out items and consumables should we have this PR updated to remove those sections?

File: #Scenes/Events/shop/0.t1104.tmp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+offset_right = 1068.0
+offset_bottom = 1064.0
+texture = ExtResource("3_6osfu")
+expand_mode = 1
+
+[node name="ShopItem" parent="." instance=ExtResource("4_bbct0")]
+offset_left = 216.0
+offset_top = 252.0
+offset_right = 216.0
+offset_bottom = 252.0
```
[the remark this answers is not in the document]
Comment (JonaLam) [reply #Scenes/Events/shop/0.t1104.tmp:37]: Is there a good way to revert this?

File: Managers/ShopManager.gd [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+func increase_removal_price() -> void:
+	card_removal_price += card_remocal_price_increase
+
+
+## Removes given card, takes money, and increases shop prices. Should only be used by the shop removal[br]
+## or potentionaly a specific event
+func remove_card_from_player_deck_with_price(card : CardBase) -> void:
+	CardManager.current_deck.erase(card)
+	InventoryManager.gold_component.lose_gold(card_removal_price)
```
[the remark this answers is not in the document]
Comment (JonaLam) [reply Managers/ShopManager.gd:22]: There is definitly something stoping this from being done somewhere, but I'll add checks in these areas as well just to be safe

File: Shop/ShopItem.gd (hunk 6 above)
Comment (JonaLam) [reply Shop/ShopItem.gd:19]: I'll make a debug function which gives a lot of gold instead, as that is easier and should do basicly the same thing

File: Shop/ShopItemCard.gd [hunk 9]
```diff
@@ -0,0 +1,16 @@
+extends ShopItem
+
+@export var cardWorld : CardWorld
+var card : CardBase
+
+func _ready() -> void:
+	var rand : RandomNumberGenerator = RandomNumberGenerator.new()
```
[the remark this answers is not in the document]
Comment (JonaLam) [reply Shop/ShopItemCard.gd:7]: I won't implement how this wil work with seeds, as I don't really know how it works, but I'll make a placeholder function for the time being

File: #Scenes/Events/shop/0.t1104.tmp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tysterman74) [reply #Scenes/Events/shop/0.t1104.tmp:37]: Here's a Stackoverflow on how to revert the file: https://stackoverflow.com/questions/18357511/git-remove-committed-file-after-push

But honestly... it's a .tmp file so I feel you could just delete this and push up that change. But I would sanity check that the shop scene is still working first after you delete and before you push up that change!

**Review by Tysterman74 — APPROVED** [review]

File: #Scenes/Events/shop/0.t1104.tmp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Turtyo) [reply #Scenes/Events/shop/0.t1104.tmp:37]: You can just delete the file and push the change, no one else uses the tmp file (as ty said, just check that it still works correctly before pushing, but I removed tmp file in another PR and it worked fine)

**Comment by Turtyo** [comment]
Yes, since we won't be using items and consumables, I would comment out / disable the items in this PR; the shop is still useful to buy torches and cards

**Review by Turtyo — CHANGES_REQUESTED** [review]
image
image
The overlay is not positioned correctly.

The proceed button doesn't open the map.

**Comment by JonaLam** [comment]
I removed the items and consumables from the shop, they're just not there, but I don't really see the point of deleting the files so I think it's fine to let them be there.

I also removed the proceed button and made it so you can move by opening the map.

**Review by Turtyo — APPROVED** [review]

**Review by Turtyo — APPROVED** [review]

**Pull request merged**